package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,579 @@
+                    while bbox_coords[i, m] < 0:
+                        for d in range(m+1):
+                            bbox_coords[i, d] += self.c_pbcbox.box[m][d]
+                    while bbox_coords[i, m] >= self.c_pbcbox.box[m][m]:
+                        for d in range(m+1):
+                            bbox_coords[i, d] -= self.c_pbcbox.box[m][d]
+        else:
+            for i in range(natoms):
+                for m in range(DIM):
+                    while bbox_coords[i, m] < 0:
+                        bbox_coords[i, m] += self.c_pbcbox.box[m][m]
+                    while bbox_coords[i, m] >= self.c_pbcbox.box[m][m]:
+                        bbox_coords[i, m] -= self.c_pbcbox.box[m][m]
+        return bbox_coords
+
+    def put_atoms_in_bbox(self, real[:,::1] coords):
+        return np.asarray(self.fast_put_atoms_in_bbox(coords))
+
+#########################
+# Neighbor Search Stuff #
+#########################
+
+cdef class NSResults(object):
+    cdef readonly real cutoff
+    cdef ns_int npairs
+
+    cdef real[:, ::1] coords  # shape: size, DIM
+    cdef real[:, ::1] searchcoords 
+    
+    cdef vector[intvec] indices_buffer
+    cdef vector[realvec] distances_buffer
+    cdef vector[ns_int] pairs_buffer
+    cdef vector[real] pair_distances_buffer
+    cdef vector[real] pair_distances2_buffer
+    
+    def __init__(self, real cutoff, real[:, ::1]coords, real[:, ::1]searchcoords):
+        self.cutoff = cutoff
+        self.coords = coords
+        self.searchcoords = searchcoords
+
+        self.npairs = 0
+     
+    cdef void add_neighbors(self, ns_int beadid_i, ns_int beadid_j, real distance2) nogil:
+        # Important: If this function returns ERROR, it means that memory allocation failed
+
+        self.pairs_buffer.push_back(beadid_i)
+        self.pairs_buffer.push_back(beadid_j)
+        self.pair_distances2_buffer.push_back(distance2)
+        self.npairs += 1
+    
+    def get_pairs(self):
+        return np.asarray(self.pairs_buffer).reshape(self.npairs, 2)
+
+    def get_pair_distances(self):
+        self.pair_distances_buffer = np.sqrt(self.pair_distances2_buffer)
+        return np.asarray(self.pair_distances_buffer)
+    
+    cdef create_buffers(self):
+        cdef ns_int i, beadid_i, beadid_j
+        cdef ns_int idx, nsearch
+        cdef real dist2
+        
+        nsearch = len(self.searchcoords)
+
+        self.indices_buffer = vector[intvec]()
+        self.distances_buffer = vector[realvec]()
+
+        # initialize rows corresponding to search
+        for i in range(nsearch):
+            self.indices_buffer.push_back(intvec())
+            self.distances_buffer.push_back(realvec())
+
+        
+        for i in range(0, 2*self.npairs, 2):
+            beadid_i = self.pairs_buffer[i]
+            beadid_j = self.pairs_buffer[i + 1]
+
+            dist2 = self.pair_distances2_buffer[i//2]
+
+            self.indices_buffer[beadid_i].push_back(beadid_j)
+
+            dist = sqrt(dist2)

```

Comment:
`dist` is untyped here, it might go to Python and be slow.  A trick I found was if you put `nogil` in the function declarations it will stop any non-C things happening which is nice sometimes